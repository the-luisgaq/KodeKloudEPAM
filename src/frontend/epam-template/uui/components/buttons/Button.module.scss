@use '../../assets/styles/index' as *;

@mixin button-disables-styles($bg-color, $border-color) {
    background-color: $bg-color;
    color: var(--uui-btn-caption-disabled);
    fill: var(--uui-btn-caption-disabled);
    border-color: $border-color;
}

// $bg-opacity we set here because a background and caption with opacity and they sum up, so we subtract one from other to get needed result
@mixin button-white-disables-styles($bg-opacity, $border-color) {
    background-color: color-mix(in srgb, var(--uui-neutral-0) $bg-opacity, transparent);
    border-color: $border-color;
    color: color-mix(in srgb, var(--uui-neutral-0) 40% , transparent);
    fill: color-mix(in srgb, var(--uui-neutral-0) 40% , transparent);

}

.root {
    // colors
    --uui-btn-bg: var(--uui-color-50);
    --uui-btn-bg-hover: var(--uui-color-60);
    --uui-btn-bg-active: var(--uui-color-70);
    --uui-btn-bg-disabled: var(--uui-neutral-30);
    --uui-btn-border: var(--uui-color-50);
    --uui-btn-border-hover: var(--uui-color-60);
    --uui-btn-border-active: var(--uui-color-70);
    --uui-btn-border-disabled: var(--uui-neutral-30);
    --uui-btn-caption: var(--uui-color-contrast);
    --uui-btn-caption-disabled: var(--uui-text-disabled);

    // fonts
    --uui-btn-font: var(--uui-font);
    --uui-btn-font-weight: var(--uui-font-weight, 600);

    // sizes
    --uui-btn-border-radius: var(--uui-border-radius);
    --uui-btn-min-height: var(--uui-size);
    --uui-btn-min-width: var(--uui-size);
    --uui-btn-border-width: var(--uui-border-width);
    --uui-btn-padding: var(--uui-horizontal-padding);
    --uui-btn-caption-padding: calc((var(--uui-btn-min-height) - var(--uui-btn-line-height) - (var(--uui-btn-border-width) * 2)) / 2) var(--uui-horizontal-gap);
    --uui-btn-icon-height: var(--uui-icon-size);
    --uui-btn-line-height: var(--uui-line-height);
    --uui-btn-font-size: var(--uui-font-size);

    //
    min-width: var(--uui-btn-min-width);
    padding-inline-start: calc(var(--uui-btn-padding) - var(--uui-btn-border-width));
    padding-inline-end: calc(var(--uui-btn-padding) - var(--uui-btn-border-width));
    background-color: var(--uui-btn-bg);
    color: var(--uui-btn-caption);
    fill: var(--uui-btn-caption);
    border-color: var(--uui-btn-border);

    &:global(.-clickable) {
        &:hover {
            background-color: var(--uui-btn-bg-hover);
            border-color: var(--uui-btn-border-hover);
        }

        &:active {
            background-color: var(--uui-btn-bg-active);
            border-color: var(--uui-btn-border-active);
        }
    }

    &:global(.uui-button-box) {
        border-style: solid;
        justify-content: center;
        border-radius: var(--uui-btn-border-radius);
        min-height: var(--uui-btn-min-height);
        border-width: var(--uui-btn-border-width);
    }

    &:focus-visible {
        @include focus-visible-effect();
    }

    :global(.uui-caption) {
        @include overflow-ellipsis();
        min-width: 0;
        text-align: center;
        flex: 0 1 auto;
        font-family: var(--uui-btn-font);
        font-weight: var(--uui-btn-font-weight, normal);
        user-select: none;
        padding: var(--uui-btn-caption-padding);
        line-height: var(--uui-btn-line-height);
        font-size: var(--uui-btn-font-size);
    }

    svg {
        height: var(--uui-btn-icon-height);
        width: inherit;
    }

    &:is(:global(.-clickable), :hover, :active, :focus):global(.uui-disabled),
    &:global(.uui-disabled) {
        &:global(.uui-fill-solid) {
            @include button-disables-styles(var(--uui-btn-bg-disabled), var(--uui-btn-border-disabled));
        }

        &:global(.uui-fill-outline) {
            @include button-disables-styles(var(--uui-surface-main), var(--uui-btn-border-disabled));
        }

        &:global(.uui-fill-none) {
            @include button-disables-styles(transparent, var(--uui-btn-border-disabled));
        }

        &:global(.uui-fill-ghost) {
            @include button-disables-styles(transparent, transparent);
        }
    }

    &:global(.uui-fill-solid) {
        --uui-btn-bg: var(--uui-color-50);
        --uui-btn-bg-hover: var(--uui-color-60);
        --uui-btn-bg-active: var(--uui-color-70);
        --uui-btn-border: var(--uui-color-50);
        --uui-btn-border-hover: var(--uui-color-60);
        --uui-btn-border-active: var(--uui-color-70);
        --uui-btn-caption: var(--uui-color-contrast);
        --uui-btn-caption-disabled: var(--uui-text-disabled);

        &:global(.uui-color-white) {
            &:active {
                background-color: var(--uui-neutral-30);
                border-color: var(--uui-neutral-30);
            }

            &:global(.uui-disabled) {
                @include button-white-disables-styles(10%, transparent);
            }
        }
    }

    &:global(.uui-fill-outline) {
        --uui-btn-bg: var(--uui-surface-main);
        --uui-btn-bg-hover: var(--uui-color-10);
        --uui-btn-bg-active: var(--uui-color-20);
        --uui-btn-bg-disabled: transparent;
        --uui-btn-border: var(--uui-color-50);
        --uui-btn-border-hover: var(--uui-color-50);
        --uui-btn-border-active: var(--uui-color-50);
        --uui-btn-caption: var(--uui-color-50);
        --uui-btn-caption-disabled: var(--uui-text-disabled);

        &:global(.uui-color-white) {
            --uui-btn-bg: var(--uui-color-5);

            &:global(.uui-disabled) {
                @include button-white-disables-styles(10%, color-mix(in srgb, var(--uui-neutral-0) 30%, transparent));
            }
        }
    }

    &:global(.uui-fill-none) {
        --uui-btn-bg: transparent;
        --uui-btn-bg-hover: var(--uui-color-10);
        --uui-btn-bg-active: var(--uui-color-20);
        --uui-btn-bg-disabled: transparent;
        --uui-btn-border: var(--uui-color-50);
        --uui-btn-border-hover: var(--uui-color-50);
        --uui-btn-border-active: var(--uui-color-50);
        --uui-btn-caption: var(--uui-color-50);
        --uui-btn-caption-disabled: var(--uui-text-disabled);

        &:global(.uui-color-white) {
            &:global(.uui-disabled) {
                @include button-white-disables-styles(0%, color-mix(in srgb, var(--uui-neutral-0) 30%, transparent));
            }
        }
    }

    &:global(.uui-fill-ghost) {
        --uui-btn-bg: transparent;
        --uui-btn-bg-hover: var(--uui-color-10);
        --uui-btn-bg-active: var(--uui-color-20);
        --uui-btn-bg-disabled: transparent;
        --uui-btn-border: transparent;
        --uui-btn-border-hover: var(--uui-color-10);
        --uui-btn-border-active: var(--uui-color-20);
        --uui-btn-border-disabled: transparent;
        --uui-btn-caption: var(--uui-color-50);
        --uui-btn-caption-disabled: var(--uui-text-disabled);

        &:global(.uui-color-white) {
            --uui-btn-border-hover: transparent;
            --uui-btn-border-active: transparent;

            &:global(.uui-disabled) {
                @include button-white-disables-styles(0%, transparent);
            }
        }
    }
}

