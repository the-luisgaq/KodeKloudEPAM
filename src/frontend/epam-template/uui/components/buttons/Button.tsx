import * as uuiComponents from '@epam/uui-components';
import type { Overwrite } from '@epam/uui-core';
import { withMods } from '@epam/uui-core';
import type { ControlSize } from '../types';
import { settings } from '../../settings';
import css from './Button.module.scss';

type ButtonMods = {
    /**
     * Defines component size.
     * @default '36'
     */
    size?: '18' | ControlSize | '60';
    /**
     * Defines component fill style.
     * @default 'solid'
     */
    fill?: 'solid' | 'outline' | 'ghost' | 'none';
    /**
     * Defines component color.
     * @default 'primary'
     */
    color?: 'accent' | 'primary' | 'critical' | 'secondary' | 'neutral' | 'white';
};
export interface ButtonModsOverride {}

/** Represents the 'Core properties' for the Button component. */
export type ButtonCoreProps = uuiComponents.ButtonProps;

/** Represents the props for a Button component. */
export interface ButtonProps extends ButtonCoreProps, Overwrite<ButtonMods, ButtonModsOverride> {}

function applyButtonMods(mods: ButtonProps) {
    return [
        css.root,
        'uui-button',
        `uui-fill-${mods.fill || 'solid'}`,
        `uui-color-${mods.color || 'primary'}`,
        `uui-size-${mods.size || settings.button.sizes.default}`,
    ];
}

export const Button = withMods<uuiComponents.ButtonProps, ButtonProps>(
    uuiComponents.Button,
    applyButtonMods,
    () => {
        return {
            dropdownIcon: settings.button.icons.dropdownIcon,
            clearIcon: settings.button.icons.clearIcon,
        };
    },
);
