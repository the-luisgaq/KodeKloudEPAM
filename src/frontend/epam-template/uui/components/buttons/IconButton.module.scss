@use '../../assets/styles/effects' as *;

.root {
    --uui-icon_btn: var(--uui-color-50);
    --uui-icon_btn-hover: var(--uui-color-60);
    --uui-icon_btn-active: var(--uui-color-70);
    --uui-icon_btn-disabled: var(--uui-icon-disabled); // from core or color class? Current - from core

    --uui-icon_btn-focus-radius: var(--uui-focus-radius);
    //
    fill: var(--uui-icon_btn);

    &:global(.uui-icon_button.uui-color-neutral) {
        --uui-icon_btn: var(--uui-icon);
        --uui-icon_btn-hover: var(--uui-icon-hover);
        --uui-icon_btn-active: var(--uui-icon-active);
    }

    &:global(.-clickable) {
        &:hover {
            fill: var(--uui-icon_btn-hover);
        }

        &:active {
            fill: var(--uui-icon_btn-active);
        }

        &:focus-visible {
            @include focus-visible-effect();
            border-radius: var(--uui-icon_btn-focus-radius);
        }
    }

    &:global(.uui-disabled) {
        fill: var(--uui-icon_btn-disabled);

        &:global(.-clickable) {
            &:hover, &:focus {
                fill: var(--uui-icon_btn-disabled);
            }

            &:active {
                fill: var(--uui-icon_btn-disabled);
            }
        }
    }
}
