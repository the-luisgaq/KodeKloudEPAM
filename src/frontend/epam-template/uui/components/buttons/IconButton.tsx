import * as uuiComponents from '@epam/uui-components';
import { withMods, Overwrite, devLogger } from '@epam/uui-core';
import { settings } from '../../settings';

import css from './IconButton.module.scss';

interface IconButtonMods {
    /**
     * Defines component color.
     * @default 'neutral'
     */
    color?: 'info' | 'success' | 'error' | 'primary' | 'accent' | 'critical' | 'warning' | 'secondary' | 'neutral' | 'white';

    /**
     * Defines component size.
     */
    size?: '18' | '24' | '30' | '36';
}

/** Represents the Core properties of the IconButton component. */
export type IconButtonCoreProps = Omit<uuiComponents.IconButtonProps, 'size'>;

export interface IconButtonModsOverride {}

/** Represents the properties of the IconButton component. */
export interface IconButtonProps extends IconButtonCoreProps, Overwrite<IconButtonMods, IconButtonModsOverride> {}

function applyIconButtonMods(props: IconButtonProps) {
    return ['uui-icon_button', `uui-color-${props.color || 'neutral'}`, css.root];
}

export const IconButton = withMods<uuiComponents.IconButtonProps, IconButtonProps>(
    uuiComponents.IconButton,
    applyIconButtonMods,
    (props) => {
        const isDeprecated = ['info', 'success', 'error', 'warning'].includes(props.color);
        if (__DEV__ && isDeprecated) {
            devLogger.warnAboutDeprecatedPropValue<IconButtonProps, 'color'>({
                component: 'IconButton',
                propName: 'color',
                propValue: props.color,
                condition: () => isDeprecated,
            });
        }
        return {
            dropdownIcon: props.dropdownIcon || settings.iconButton.icons.dropdownIcon,
        };
    },
);
