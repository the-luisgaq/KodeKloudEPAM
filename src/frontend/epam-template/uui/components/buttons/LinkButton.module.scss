@use  '../../assets/styles/index' as *;

.root {
    // colors
    --uui-link_btn-text: var(--uui-color-50);
    --uui-link_btn-text-hover: var(--uui-color-60);
    --uui-link_btn-text-active: var(--uui-color-70);
    --uui-link_btn-text-disabled: var(--uui-text-disabled);

    // fonts
    --uui-link-button-font: var(--uui-font);
    --uui-link-button-font-weight-semibold: 600;
    --uui-link-button-font-weight-regular: 400;

    // sizes
    --uui-link-button-min-height: var(--uui-size);
    --uui-link-button-min-width: var(--uui-size);
    --uui-link-button-gap: var(--uui-horizontal-gap);
    --uui-link-button-caption-padding: calc((var(--uui-link-button-min-height) - var(--uui-link-button-line-height)) / 2);
    --uui-link-button-icon-height: var(--uui-icon-size);
    --uui-link-button-line-height: var(--uui-line-height);
    --uui-link-button-font-size: var(--uui-font-size);
    --uui-link-button-focus-radius: var(--uui-focus-radius);
    //
    min-width: var(--uui-link-button-min-width);

    @include clickable-styles();

    &:global(.uui-button-box) {
        background-color: transparent;
        color: var(--uui-link_btn-text);
        fill: var(--uui-link_btn-text);
        min-height: var(--uui-link-button-min-height);
        min-width: auto;

        &:global(.-clickable:not(.uui-disabled)) {
            &:hover {
                color: var(--uui-link_btn-text-hover);
                fill: var(--uui-link_btn-text-hover);
            }

            &:active {
                color: var(--uui-link_btn-text-active);
                fill: var(--uui-link_btn-text-active);
            }

            &:focus {
                box-shadow: none;
            }

            &:focus-visible {
                @include focus-visible-effect(2px);
                border-radius: var(--uui-link-button-focus-radius);
            }
        }

        &:global(.uui-disabled) {
            color: var(--uui-link_btn-text-disabled);
            fill: var(--uui-link_btn-text-disabled);
            background-color: transparent;

            &:hover {
                background-color: transparent;
            }
        }
    }

    :global(.uui-caption) {
        font-family: var(--uui-link-button-font);
        font-weight: var(--uui-link-button-font-weight, normal); // --uui-link-button-font-weight - don't delete value specified inline
        line-height: var(--uui-link-button-line-height);
        font-size: var(--uui-link-button-font-size);
        padding-top: var(--uui-link-button-caption-padding);
        padding-bottom: var(--uui-link-button-caption-padding);

        &:not(:first-child) {
            padding-inline-start: var(--uui-link-button-gap);
        }

        &:not(:last-child) {
            padding-inline-end: var(--uui-link-button-gap);
        }

        &:global(.uui-underline-solid) {
            text-decoration: underline;
            text-underline-offset: 0.2em;
        }

        &:global(.uui-underline-dashed) {
            text-decoration: underline dashed;
            text-underline-offset: 0.2em;
        }

        &:global(.uui-link-button-weight-semibold) {
            font-weight: var(--uui-link-button-font-weight-semibold);
        }

        &:global(.uui-link-button-weight-regular) {
            font-weight: var(--uui-link-button-font-weight-regular);
        }
    }

    :global(.uui-icon) {
        svg {
            height: var(--uui-link-button-icon-height);
            width: inherit;
        }
    }
}
