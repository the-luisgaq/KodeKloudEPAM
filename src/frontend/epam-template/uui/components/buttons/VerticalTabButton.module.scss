@use '../../assets/styles/index' as *;

.root {
    border-bottom: 0 none;
    border-inline-start: var(--uui-tab-line-width) solid transparent;
    text-align: start;
    column-gap: 0;
    padding-inline-start: calc(var(--uui-tab_btn-padding) - var(--uui-tab-line-width));

    :global(.uui-caption) {
        &:first-child {
            padding-inline-start: 0;
        }

        &:last-child {
            padding-inline-end: 0;
        }
    }

    :global(.uui-count_indicator) {
        margin-inline-start: 3px;
        margin-inline-end: var(--uui-horizontal-gap);

        ~ :global(.uui-caption) {
            padding-inline-end: 0;
        }
    }
}
