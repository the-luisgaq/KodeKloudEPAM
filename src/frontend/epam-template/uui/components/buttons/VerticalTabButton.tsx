import { withMods } from '@epam/uui-core';
import { TabButton, TabButtonProps } from './TabButton';
import css from './VerticalTabButton.module.scss';

function applyVerticalTabButtonMods() {
    return [css.root, 'uui-vertical-tab-button'];
}

/** Represents the properties of a VerticalTabButton component. */
export type VerticalTabButtonProps = TabButtonProps;

export const VerticalTabButton = withMods<VerticalTabButtonProps, VerticalTabButtonProps>(TabButton, applyVerticalTabButtonMods);
