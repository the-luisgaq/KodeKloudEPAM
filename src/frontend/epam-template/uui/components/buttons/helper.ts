import { ButtonProps } from '@epam/uui-components';

export function getIconClass(props: ButtonProps) {
    const classList = {
        'has-left-icon': false,
        'has-right-icon': false,
    };

    if (props.onClear) {
        classList['has-right-icon'] = true;
    }

    if (props.isDropdown) {
        classList['has-right-icon'] = true;
    }

    if (props.icon) {
        classList[props.iconPosition !== 'right' ? 'has-left-icon' : 'has-right-icon'] = true;
    }

    return [classList['has-left-icon'] ? 'uui-has-left-icon' : 'uui-no-left-icon', classList['has-right-icon'] ? 'uui-has-right-icon' : 'uui-no-right-icon'];
}
