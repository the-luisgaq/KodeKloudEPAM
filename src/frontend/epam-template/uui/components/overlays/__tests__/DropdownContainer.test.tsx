import React from 'react';
import { DropdownContainer } from '../DropdownContainer';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { Button } from '../../buttons';

describe('DropdownContainer', () => {
    it('should be rendered correctly with minimum props', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <DropdownContainer>
                <Button />
            </DropdownContainer>,
        );
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <DropdownContainer width={ 300 } height={ 600 } vPadding="18" padding="24" focusLock={ false }>
                <Button />
            </DropdownContainer>,
        );
        expect(tree).toMatchSnapshot();
    });
});
