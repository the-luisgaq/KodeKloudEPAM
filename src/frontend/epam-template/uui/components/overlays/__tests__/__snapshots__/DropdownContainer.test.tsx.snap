// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownContainer should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="uui-dropdown-body -lock-focus root vPadding-18 padding-24 container"
    style="min-width: 300px; min-height: 600px; max-width: 300px;"
    tabindex="-1"
  >
    <button
      class="uui-button-box uui-enabled container root uui-button uui-fill-solid uui-color-primary uui-size-36"
      tabindex="-1"
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`DropdownContainer should be rendered correctly with minimum props 1`] = `
<DocumentFragment>
  <div
    class="uui-dropdown-body -lock-focus root container"
    tabindex="-1"
  >
    <button
      class="uui-button-box uui-enabled container root uui-button uui-fill-solid uui-color-primary uui-size-36"
      tabindex="-1"
      type="button"
    />
  </div>
</DocumentFragment>
`;
